mmaprototype: add tracing to testing for rebalanceStores

tbg opened the pull request:
This PR adds the basic capability to exercise rebalanceStores through
TestClusterState. It does this by moving rebalanceStores from allocatorState to
clusterState, which is a small and mostly mechanical refactor.

The immediate motivation behind this change is being able to exercise
the functionality for #156776 while it is being added, while being able
to see the internal tracepoints as well.

Informs #156776.

Epic: CRDB-55052

tbg committed:
mmaprototype: add tracing to rebalance_store test

tbg committed:
tracing: add minimal format suitable for datadriven tests

I'd like to make the trace events emitted from deterministic multi-metric
allocator tests part of the expected test outputs.

I realized there wasn't a standardized deterministic and barebones output
format, so I added one.

Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/157133)
<!-- Reviewable:end -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> @sumeerbhola reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157133#-:-Odj_19VEvrUezRFqMlP:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @arjunmahishi, @jasonlmfong, and @wenyihu6)



<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (approved):
mod the comment 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157133#-:-Odj_eMsCZjPoluUVNvD:bll9jgz)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @arjunmahishi, and @jasonlmfong)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_test.go` line 524 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157133#-Odj_UiuDd7uHJcU6RNu:-Odj_Uiv0lkgk_FI8_Zb:bkd766j) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8528574861198f9a80e3994324a38f1b2e65062/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_test.go#L524)):*
> ```Go
> 					rng := rand.New(rand.NewSource(0))
> 					dsm := newDiversityScoringMemo()
> 					tr := tracing.NewTracer()
> ```

do we need to call tr.Close in the end like this https://github.com/cockroachdb/cockroach/blob/ba2a5d4f6f7a31ceec030692d3e0eb2e79380173/pkg/kv/kvserver/kvstorage/datadriven_test.go#L116? 


<!-- Sent from Reviewable.io -->

tbg committed:
fixup! mmaprototype: add tracing to rebalance_store test

close tracer, ./dev gen bazel, set tracer redactable (make sure we get good debug output in production)

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157133)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aa-joshi, @Abhinav1299, @arjunmahishi, @jasonlmfong, @sumeerbhola, and @wenyihu6)

___
*[`pkg/kv/kvserver/allocator/mmaprototype/cluster_state_test.go` line 524 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/157133#-Odj_UiuDd7uHJcU6RNu:-OdqwGlB9PBGi5sSuw-O:bxzyohp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8528574861198f9a80e3994324a38f1b2e65062/pkg/kv/kvserver/allocator/mmaprototype/cluster_state_test.go#L524)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

do we need to call tr.Close in the end like this https://github.com/cockroachdb/cockroach/blob/ba2a5d4f6f7a31ceec030692d3e0eb2e79380173/pkg/kv/kvserver/kvstorage/datadriven_test.go#L116?
</blockquote></details>

We should, thanks. Added.


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157133#-:-Odsa7s74XCoGrMIhIEm:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @aa-joshi, @Abhinav1299, @arjunmahishi, @jasonlmfong, and @sumeerbhola)



<!-- Sent from Reviewable.io -->

tbg commented:
Ugh, wanted to unstack this before merging and squash the fixup commit. Making a mental note to add do-not-merge label when in that state.
